engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingformula/BaseRollingFormulaOperator.java

```diff
@@ -0,0 +1,291 @@
+
+        // Re-use the formula column if it's already been created for this type. No need to synchronize; these
+        // operators are created serially.
+        // TODO: does generic Object need to be handled uniquely?
+        formulaColumn = formulaColumnMap.computeIfAbsent(columnType, t -> {
+            final FormulaColumn tmp = FormulaColumn.createFormulaColumn(outputColumnName,
+                    FormulaUtil.replaceFormulaTokens(formulaToUse, paramToken, PARAM_COLUMN_NAME));
+
+            final ColumnDefinition<?> inputColumnDefinition = ColumnDefinition
+                    .fromGenericType(PARAM_COLUMN_NAME, vectorType, columnType);
+            tmp.initDef(Collections.singletonMap(PARAM_COLUMN_NAME, inputColumnDefinition));
+            return tmp;
+        });
+    }
+
+    protected BaseRollingFormulaOperator(
+            @NotNull final MatchPair pair,
+            @NotNull final String[] affectingColumns,
+            @Nullable final String timestampColumnName,
+            final long reverseWindowScaleUnits,
+            final long forwardWindowScaleUnits,
+            final Class<?> vectorType,
+            @NotNull final Map<Class<?>, FormulaColumn> formulaColumnMap,
+            @NotNull final TableDefinition tableDef) {
+        super(pair, affectingColumns, timestampColumnName, reverseWindowScaleUnits, forwardWindowScaleUnits, true);
+        this.formulaColumnMap = formulaColumnMap;
+        this.tableDef = tableDef;
+
+        final Class<?> columnType = tableDef.getColumn(pair.rightColumn).getDataType();
+        this.vectorType = vectorType;
+
+        // Re-use the formula column that already been created for this type.
+        // TODO: does generic Object need to be handled uniquely?
+        formulaColumn = formulaColumnMap.computeIfAbsent(columnType, t -> {
+            throw new IllegalStateException("formulaColumnMap should have been populated for " + columnType);
+        });
+    }
+
+    @Override
+    public void initializeSources(@NotNull final Table source, @Nullable final RowRedirection rowRedirection) {
+        this.rowRedirection = rowRedirection;
+
+        if (rowRedirection != null) {
+            // region create-dense
+            maybeInnerSource = ArrayBackedColumnSource.getMemoryColumnSource(0, formulaColumn.getReturnedType());
+            // endregion create-dense
+            outputSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, maybeInnerSource, 0);
+        } else {
+            maybeInnerSource = null;
+            // region create-sparse
+            outputSource = SparseArrayColumnSource.getSparseMemoryColumnSource(0, formulaColumn.getReturnedType());
+            // endregion create-sparse
+        }
+        outputChunkType = outputSource.getChunkType();
+        nullValue = getNullValue(outputChunkType);
+    }
+
+    private static Class<?> getVectorType(final Class<?> type) {
```

Comment:
Note that `DhFormulaColumn#getVectorType` checks for boxed primitives; you may want to ensure that `Long` source columns still use `LongRollingFormulaOperator` and not the `Object` impl.